File: mindflow/src/utils/files/extract.rs
use std::{path::Path, fs};

use super::git::{is_within_git_repo, get_git_files};



pub fn extract_files(path: &Path) -> Vec<String> {
    // If the path is a file, return it as the only item in a vector
    if path.is_file() {
        return vec![path.to_string_lossy().to_string()];
    }

    // Check if the path is within a git repository
    if is_within_git_repo(path) {
        return get_git_files(path);
    } 

    // Read all the files and directories in the path, and recursively extract files from directories
    let mut file_paths = Vec::new();
    match fs::read_dir(path) {
        Ok(entries) => {
            for entry in entries {
                match entry {
                    Ok(entry) => {
                        let path = entry.path();
                        match path.is_dir() {
                            true => {
                                file_paths.extend(extract_files(&path));
                            },
                            false => {
                                file_paths.push(path.to_string_lossy().to_string());
                            }
                        }
                    },
                    // If the directory could not be read, log the error
                    Err(e) => {
                        log::debug!("Could not read directory: {}. Error: {:?}", path.to_string_lossy(), e);
                    }
                }
            }
            file_paths
        },
        // If the directory could not be read, log the error
        Err(e) => {
            log::debug!("Could not read directory: {}. Error: {:?}", path.to_string_lossy(), e);
            file_paths
        }
    }
}

File: mindflow/utils/files/extract.py
"""
Module for recursively extracting all files from a file path.
"""

import logging
import os

from typing import List, Union

from mindflow.utils.files.git import is_within_git_repo, get_git_files


def extract_files(path: Union[str, os.PathLike]) -> List[str]:
    """
    Extract all files from a directory.
    """
    file_paths = []
    if os.path.isfile(path):
        return [os.fspath(path)]
    if is_within_git_repo(path):
        return get_git_files(path)
    try:
        with os.scandir(path) as paths:
            for scanned_path in paths:
                if scanned_path.is_file():
                    file_paths.append(os.path.abspath(scanned_path.path))
                elif os.path.isdir(path):
                    file_paths.extend(extract_files(scanned_path.path))
        return file_paths
    except IsADirectoryError as error:
        logging.debug("Could not read directory/file: %s", path)
        logging.debug(error)
        return file_paths
